### DISCUSSION TITLE
Is the Start Debugging button on the editor toolbar necessary?

### DISCUSSION THREAD

**Discussion by zero-plusplus** [zero-plusplus/vscode-autohotkey-debug#117] [Ideas]
The editor toolbar is number 10 in the image below.
![ui](https://code.visualstudio.com/assets/api/references/guidelines/architecture-sections.png)

Since I have not used this UI before, I did not add a start debugging button as seen in other extensions.

But, it looks like the next version will probably be able to use drop-down buttons, which will improve usability, so I wanted to take this opportunity to open a discussion to ask for your opinions.

Note that this feature will not be implemented immediately.
The reason for this is that if I incorporate the features of the latest version of vscode, the extension will not work on earlier versions.

**Comment by SAbboushi** [comment]
I use 2 monitors for vscode debugging.
Main monitor:
![image](https://user-images.githubusercontent.com/1964198/130308757-a273631d-fa9b-406b-ab9b-44a77e618115.png)
Secondary monitor:
![image](https://user-images.githubusercontent.com/1964198/130308770-af8644f3-04ff-46ab-abe1-42420c26cf8e.png)

I think I would prefer start debugging button/dropdown and debug toolbar to be in the Editor Toolbar which is above the code I am debugging instead of having to control from second monitor.  But this is not a priority for me as long as I can continue to drag the debug toolbar to the Editor Toolbar area:
![image](https://user-images.githubusercontent.com/1964198/130308967-f82db627-023b-43d9-823c-c1452e7af8bb.png)

**Reply by zero-plusplus** [reply]
Thanks for the feedback.

It's true that I don't need to bother with this button if I can use the debug toolbar.

What I find inconvenient with the debug toolbar is that when I am using a configuration that debugs a specific file, I cannot debug the file that is open. If I want to do that, I have to switch configurations.

So, how about implementing it as a function to debug the currently open file?
This will differentiate it from the debug toolbar and reduce the need to switch between debug configurations.

If I can use a drop-down button, I can put together the two functions of `run/debug the currently open file`.

**Comment by SAbboushi** [comment]
>> What I find inconvenient with the debug toolbar is that when I am using a configuration that debugs a specific file, I cannot debug the file that is open. If I want to do that, I have to switch configurations.

I recall that our configurations are very different: I'm not sure I understand what you mean.  Can you explain more please?

**Reply by zero-plusplus** [reply]
For example, let's consider the following case where the main script is the target of debugging.
```
...
{
    "name": "AutoHotkey Debug",
    "request": "launch",
    "type": "autohotkey",
    "runtime": "${workspaceFolder}/AutoHotkey.exe",
    "program": "${workspaceFolder}/AutoHotkey.ahk",
},
...
```

However, when debugging each library used in the main script, it is more convenient to use the open file as the debug target, as shown below.
```
...
{
    "name": "AutoHotkey Debug (Selected File)",
    "request": "launch",
    "type": "autohotkey",
    "runtime": "${workspaceFolder}/AutoHotkey.exe",
    "program": "${file}",
},
...
```

I find it very annoying to switch between the two. I've devised hotkeys to reduce the hassle, but it's still a pain.

The only difference between these configurations is `program`, and if there is a way to rewrite it and run debugging, we can start debugging the open file without switching configurations.

**Comment by SAbboushi** [comment]
Since my vscode configuration is very different (at least it was in the past when we dialoged about it), I'm having difficulty visualizing your workflow for when you want to use the main script for debugging. If you don't mind sharing with me, what are the steps when you want to debug the main script?  Is vscode already open?  Is the main script already opened in the Editor?

**Reply by zero-plusplus** [reply]
The file structure of my project is as follows.
```
root
  - lib
    -  Util-A.ahk
    -  Util-B.ahk
  - test
    -  Util-A_test.ahk
    -  Util-B_test.ahk
  - AutoHotkey.ahk
  - AutoHotkey.exe
  - HotkeyDefinition_VSCode.ahk
  - HotkeyDefinition_GoogleChrome.ahk
```

The file starting with `HotkeyDefinition` is a subfile of the main script `AutoHotkey.ahk` and will cause an error if run by itself.
Therefore, it is usually only `AutoHotkey.ahk` that is targeted for debugging.

However, if I want to debug `Util-A_test.ahk`, which is a unit test of the library, I need to switch to a different debugging configuration.

Since unit tests usually allow each file to be executed separately, it is more efficient to use the open file as the debug target rather than creating a separate debug configuration for each.

**Comment by SAbboushi** [comment]
I'm guessing: your workflow is
1) You launch vscode
2) Your default configuration is for AutoHotkey.ahk, so you can just press F5 key or click the green triangle, although I suspect yours says "AutoHotkey Debug" and doesn't say "(workspace)":
![image](https://user-images.githubusercontent.com/1964198/130362314-fd84b73e-9271-4267-9b12-75a2f6bd14f7.png)

Then when you want to debug a different file, you open the file in vscode and then click on the dropdown to select "AutoHotkey Debug (Selected File)" and then press F5 key or click the green start triangle.

>> So, how about implementing it as a function to debug the currently open file?
Now it is clear to me what you mean.  If I understand correctly, you then will no longer need the "AutoHotkey Debug (Selected File)" configuration because you can open any file and then click on the Editor Toolbar Start button.  

My thoughts, assuming that you have implemented a "Start Debugging" button on the editor toolbar: 
I think your idea is very useful since I suspect it is a common need to debug the currently opened file.
It seems that if the debugger is not running, the F5 key would start the debugger using the the current configuration (i.e. the one displayed in the "RUN AND DEBUG" header of the "[Debug side bar](https://code.visualstudio.com/docs/editor/debugging)"?
And if the debugger is already running, then the shortcut keys (F5, F10, F11...) would function on the active debug session regardless of whether it was started using the "Start Debugging" button on the editor toolbar or the "Start Debugging" button on the Debug side bar?
What configuration would be used for the "Start Debugging" button on the editor toolbar (e.g. the "runtime" parameter)?

**Reply by zero-plusplus** [reply]
You guessed it right.

> What configuration would be used for the "Start Debugging" button on the editor toolbar (e.g. the "runtime" parameter)?

There are three options for this.
1. override the currently selected debug configuration.
2. add a option for the landmark to launch.json and use that configuration
3. add the name of the debug configuration you want to use to settings.json and use that configuration

I think 3 is the easiest to implement and understand.

**Comment by SAbboushi** [comment]
>> I think 3 is the easiest to implement and understand
I think that makes sense too

But, my debug configurations are specified in Workspace Settings (e.g. "AHKv2.code-workspace"), not User Settings (settings.json).  When we dialoged about configurations before, I think maybe you were not using Workspace Settings.  I hope that your extension and features such as this will continue to work with Workspace Settings.

**Reply by zero-plusplus** [reply]
I forgot about code-workspace because I don't use it regularly.
Don't worry, I will implement it so that you can use it even if you are using code-workspace.

---

My extension does not take code-workspace into account, so it may contain inherent issues. If you encounter any issues, such as the inability to use a feature, please report them to me.

**Comment by SAbboushi** [comment]
Thank you
